### PR TITLE
Improving the error details for hand shake issues during brokered authentication on android.

### PR DESCRIPTION
**Pull request by trwalke**
Fix for https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/2045

### CODE REVIEW

**Review by neha-bhargava — APPROVED** [review]
LGTM

File: src/client/Microsoft.Identity.Client/Platforms/Android/Broker/AndroidBrokerHelper.cs [hunk 1]
```diff
@@ -318,7 +318,27 @@ public async Task InitiateBrokerHandshakeAsync(Activity callerActivity)
                             return;
                         }
 
-                        throw new MsalClientException("Could not negotiate protocol version with broker.");
+                        dynamic errorResult = JObject.Parse(bundleResult?.GetString(BrokerConstants.BrokerResultV2));
+                        string error = null;
+                        string errorDescription = null;
+
+                        if (errorResult != null)
```
Comment (henrik-me) [inline src/client/Microsoft.Identity.Client/Platforms/Android/Broker/AndroidBrokerHelper.cs:325]: >errorResult != null [](start = 28, length = 19)

string.isnullorempty??

File: src/client/Microsoft.Identity.Client/Platforms/Android/Broker/AndroidBrokerHelper.cs [hunk 2]
```diff
@@ -318,7 +318,27 @@ public async Task InitiateBrokerHandshakeAsync(Activity callerActivity)
...
+                            errorDescription = errorResult[BrokerResponseConst.BrokerErrorMessage]?.ToString();
+
+                            _logger.Error($"error: {error} errorDescription {errorDescription}");
+
+                        }
+                        else
+                        {
+                            error = BrokerConstants.BrokerUnknownErrorCode;
+                            errorDescription = "Error Code received during hand shake, but no error could be extracted";
+                            _logger.Error("Error response received during hand shake, but not error could be extracted");
```
Comment (henrik-me) [inline src/client/Microsoft.Identity.Client/Platforms/Android/Broker/AndroidBrokerHelper.cs:337]: >not [](start = 90, length = 3)

remove t

File: src/client/Microsoft.Identity.Client/Platforms/Android/Broker/AndroidBrokerHelper.cs [hunk 3]
```diff
@@ -318,7 +318,27 @@ public async Task InitiateBrokerHandshakeAsync(Activity callerActivity)
                             return;
                         }
 
-                        throw new MsalClientException("Could not negotiate protocol version with broker.");
+                        dynamic errorResult = JObject.Parse(bundleResult?.GetString(BrokerConstants.BrokerResultV2));
+                        string error = null;
```
Comment (henrik-me) [inline src/client/Microsoft.Identity.Client/Platforms/Android/Broker/AndroidBrokerHelper.cs:322]: nit: errorCode for clarity

File: src/client/Microsoft.Identity.Client/Platforms/Android/Broker/AndroidBrokerHelper.cs [hunk 4]
```diff
@@ -318,7 +318,27 @@ public async Task InitiateBrokerHandshakeAsync(Activity callerActivity)
...
+                        string errorDescription = null;
+
+                        if (errorResult != null)
+                        {
+                            error = errorResult[BrokerResponseConst.BrokerErrorCode]?.ToString();
+                            errorDescription = errorResult[BrokerResponseConst.BrokerErrorMessage]?.ToString();
+
+                            _logger.Error($"error: {error} errorDescription {errorDescription}");
```
Comment (henrik-me) [inline src/client/Microsoft.Identity.Client/Platforms/Android/Broker/AndroidBrokerHelper.cs:330]: >e [](start = 44, length = 1)

nit: capitalize E in error and errorDescription

Comment (jennyf19) [reply src/client/Microsoft.Identity.Client/Platforms/Android/Broker/AndroidBrokerHelper.cs]: does the error message from broker contain any PII? 

---
In reply to: [490529201](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2047#discussion_r490529201) [](ancestors = 490529201)

File: src/client/Microsoft.Identity.Client/Platforms/Android/Broker/AndroidBrokerHelper.cs [hunk 5]
```diff
@@ -318,7 +318,28 @@ public async Task InitiateBrokerHandshakeAsync(Activity callerActivity)
...
+                        else
+                        {
+                            errorCode = BrokerConstants.BrokerUnknownErrorCode;
+                            errorDescription = "An error occurred during hand shake with the broker, no detailed error information was returned";
+
+                            _logger.Error(errorDescription);
+                        }
+
+                        throw new MsalClientException(errorCode, 
+                            "Could not negotiate protocol version with broker due to the following error: " + errorDescription);
```
Comment (henrik-me) [inline src/client/Microsoft.Identity.Client/Platforms/Android/Broker/AndroidBrokerHelper.cs:342]: >"Could not negotiate protocol version with broker due to the following error: " [](start = 28, length = 79)

is this needed? Why not "just" the error description you already collected above?

Comment (trwalke) [reply src/client/Microsoft.Identity.Client/Platforms/Android/Broker/AndroidBrokerHelper.cs]: you are right. that info is not needed.

File: src/client/Microsoft.Identity.Client/Platforms/Android/Broker/AndroidBrokerHelper.cs [hunk 6]
```diff
@@ -318,7 +318,27 @@ public async Task InitiateBrokerHandshakeAsync(Activity callerActivity)
...
 
-                        throw new MsalClientException("Could not negotiate protocol version with broker.");
+                        dynamic errorResult = JObject.Parse(bundleResult?.GetString(BrokerConstants.BrokerResultV2));
+                        string errorCode = null;
+                        string errorDescription = null;
+
+                        if (!string.IsNullOrEmpty(errorResult))
+                        {
+                            errorCode = errorResult[BrokerResponseConst.BrokerErrorCode]?.ToString();
+                            errorDescription = errorResult[BrokerResponseConst.BrokerErrorMessage]?.ToString();
```
Comment (henrik-me) [inline src/client/Microsoft.Identity.Client/Platforms/Android/Broker/AndroidBrokerHelper.cs:328]: >errorDescription [](start = 28, length = 16)

nit: wouldn't this be the error Message, and the below thing you log be the error description?

File: src/client/Microsoft.Identity.Client/Platforms/Android/Broker/AndroidBrokerHelper.cs [hunk 7]
```diff
@@ -318,7 +318,27 @@ public async Task InitiateBrokerHandshakeAsync(Activity callerActivity)
...
+                        dynamic errorResult = JObject.Parse(bundleResult?.GetString(BrokerConstants.BrokerResultV2));
+                        string errorCode = null;
+                        string errorMessage = null;
+
+                        if (!string.IsNullOrEmpty(errorResult))
+                        {
+                            errorCode = errorResult[BrokerResponseConst.BrokerErrorCode]?.ToString();
+                            errorMessage = errorResult[BrokerResponseConst.BrokerErrorMessage]?.ToString();
+
+                            _logger.Error($"An error occurred during hand shake with the broker. Error: {errorCode} Error Message: {errorMessage}");
```
Comment (henrik-me) [inline src/client/Microsoft.Identity.Client/Platforms/Android/Broker/AndroidBrokerHelper.cs:330]: >$"An error occurred during hand shake with the broker. Error: {errorCode} Error Message: {errorMessage}" [](start = 42, length = 104)

this is the errorDescription

File: src/client/Microsoft.Identity.Client/Platforms/Android/Broker/AndroidBrokerHelper.cs [hunk 8]
```diff
@@ -318,7 +318,27 @@ public async Task InitiateBrokerHandshakeAsync(Activity callerActivity)
...
+
+                        }
+                        else
+                        {
+                            errorCode = BrokerConstants.BrokerUnknownErrorCode;
+                            errorMessage = "An error occurred during hand shake with the broker, no detailed error information was returned";
```
Comment (henrik-me) [inline src/client/Microsoft.Identity.Client/Platforms/Android/Broker/AndroidBrokerHelper.cs:336]: >errorMessage [](start = 28, length = 12)

this is the errorDescription

File: src/client/Microsoft.Identity.Client/Platforms/Android/Broker/AndroidBrokerHelper.cs [hunk 9]
```diff
@@ -318,7 +318,27 @@ public async Task InitiateBrokerHandshakeAsync(Activity callerActivity)
...
+
+                            _logger.Error(errorMessage);
+                        }
+
+                        throw new MsalClientException(errorCode, errorMessage);
```
Comment (henrik-me) [inline src/client/Microsoft.Identity.Client/Platforms/Android/Broker/AndroidBrokerHelper.cs:341]: >errorMessage [](start = 65, length = 12)

this should be the errorDescription

**Pull request merged**